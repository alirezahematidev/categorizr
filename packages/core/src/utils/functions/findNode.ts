import { MaybeNode, TreeNode } from "$core/types";

export function findNode<T extends TreeNode, const Id extends T["id"]>(tree: readonly T[], id: Id): MaybeNode<T> {
  for (const node of tree) {
    if (node.id === id) return node;

    const foundNode = findNode<T, Id>(node.children as T[], id);

    if (foundNode) return foundNode;
  }
  return undefined;
}
