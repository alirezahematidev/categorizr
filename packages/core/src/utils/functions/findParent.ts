import { MaybeNode, TreeNode } from "$core/types";
import { noChildren } from "../helpers";

export function findParent<T extends TreeNode>(tree: readonly T[], node: T): MaybeNode<T> {
  for (const branch of tree) {
    if (noChildren(branch)) continue;

    if (branch.children.some((child) => child.id === node.id)) return branch;

    const parent = findParent<T>(branch.children as T[], node);

    if (parent) return parent;
  }

  return undefined;
}
