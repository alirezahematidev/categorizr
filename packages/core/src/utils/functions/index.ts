export * from "./findNode";
export * from "./findParent";
export * from "./insertNode";
