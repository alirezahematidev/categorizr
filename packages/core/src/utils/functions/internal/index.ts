export * from "./findNode";
export * from "./findParent";
export * from "./containsNode";
