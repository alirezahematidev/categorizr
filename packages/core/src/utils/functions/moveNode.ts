import { todo } from "../helpers";

function moveNode() {
  todo("NO IMPLEMENTATION");
}

export { moveNode };
