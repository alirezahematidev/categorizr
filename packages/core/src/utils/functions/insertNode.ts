import { TreeNode } from "$core/types";
import { findNode } from "./findNode";

function insertNode<T extends TreeNode, const Id extends T["id"]>(tree: T[], destId: Id, node: T, callback: (tree: T[]) => void): void;
function insertNode<T extends TreeNode, const Id extends T["id"]>(tree: T[], destId: Id, node: T): T[];
function insertNode(tree: any[], destId: string, node: any, callback?: (tree: any[]) => void) {
  const cloneTree = [...tree];

  const parentNode = findNode(cloneTree, destId);

  if (!parentNode) throw new Error("cannot found the parent node within given parentId.");

  if (!parentNode.children) parentNode.children = [];

  parentNode.children.push(node);

  if (callback) return void callback(cloneTree);

  return cloneTree;
}

export { insertNode };
