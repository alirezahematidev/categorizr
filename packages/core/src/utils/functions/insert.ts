import { DestId, Callback, TreeNode } from "$core/types";
import { clone, exception } from "../helpers";
import { findNode } from "./internal";

function insert<T extends TreeNode>(tree: readonly T[], destId: DestId, node: T): T[];
function insert<T extends TreeNode>(tree: readonly T[], destId: DestId, node: T, callback: Callback<T>): void;
function insert<T extends TreeNode>(tree: readonly T[], destId: DestId, node: T, callback?: Callback<T>) {
  const cloneTree = clone(tree);

  if (destId === null) {
    cloneTree.push(node);

    return cloneTree;
  }

  const parentNode = findNode(cloneTree, destId);

  if (!parentNode) throw exception("insert", "cannot found the parent node within given destId.");

  if (!parentNode.children) parentNode.children = [];

  parentNode.children.push(node);

  if (callback) return void callback(cloneTree);

  return cloneTree;
}

export { insert };
