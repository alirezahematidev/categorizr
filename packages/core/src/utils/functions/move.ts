import { insert } from "./insert";
import { remove } from "./remove";
import { Callback, DestId, TreeNode } from "$core/types";
import { containsNode, findNode, findParent } from "./internal";
import { exception } from "../helpers";

function move<T extends TreeNode>(tree: readonly T[], sourceId: string, destId: DestId): T[];
function move<T extends TreeNode>(tree: readonly T[], sourceId: string, destId: DestId, callback: Callback<T>): void;
function move<T extends TreeNode>(tree: readonly T[], sourceId: string, destId: DestId, callback?: Callback<T>) {
  const node = findNode(tree, sourceId);

  if (containsNode(tree, sourceId, destId)) throw exception("move", "Cannot move the node into its own descendants.");

  if (!node) throw exception("move", "cannot found the node within given id");

  const parent = findParent(tree, node);

  if (parent && parent.id === destId) {
    if (callback) return void callback(tree as T[]);

    return tree as T[];
  }

  const result = insert(remove(tree, sourceId), destId, node);

  if (callback) return void callback(result);

  return result;
}

export { move };
