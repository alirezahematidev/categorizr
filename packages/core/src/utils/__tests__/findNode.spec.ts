import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import type { TreeNode } from "$core/types";
import { findNode } from "../functions";

describe("findNode", async () => {
  beforeAll(() => {
    vi.fn().mockImplementation(findNode);
  });

  afterAll(() => {
    vi.resetAllMocks();
  });

  const { default: CATEGORIES } = await vi.importActual<{ default: TreeNode[] }>("$core/__mocks__");

  it("returns undefined for empty tree", () => {
    const emptyTree: TreeNode[] = [];

    const nodeId = "1";

    expect(findNode(emptyTree, nodeId)).toBeUndefined();
  });

  it("returns correct node for existing node in the tree", async () => {
    const nodeId = "4";

    const expectedNode = {
      id: "4",
      name: "sub-category-2",
      children: [],
    };

    expect(findNode(CATEGORIES, nodeId)).toStrictEqual(expectedNode);
    expect(findNode(CATEGORIES, nodeId)).toMatchSnapshot();
  });
});
