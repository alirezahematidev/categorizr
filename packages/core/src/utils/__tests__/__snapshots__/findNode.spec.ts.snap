// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`findNode > returns correct node for existing node in the tree 1`] = `
{
  "children": [],
  "id": "4",
  "name": "sub-category-2",
}
`;
